File: leetcode/golang/LCR_191_gou_jian_cheng_ji_shu_zu_lcof.go
func statisticalResult(arrayA []int) []int {
    if len(arrayA) == 0 {
        return []int{}
    }
    res := make([]int, len(arrayA))
    res[0] = 1
    for i := 1; i < len(arrayA); i++ {
        res[i] = res[i-1] * arrayA[i-1]
    }
    right := 1
    for i := len(arrayA) - 1; i >= 0; i-- {
        res[i] *= right
        right *= arrayA[i] 
    }
    return res
}

File: leetcode/c++/LCR_191_gou_jian_cheng_ji_shu_zu_lcof.cpp
// https://leetcode.cn/problems/gou-jian-cheng-ji-shu-zu-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> statisticalResult(vector<int>& arrayA) {
        if (arrayA.size() == 0)
            return {};
        vector<int> left(arrayA.size(), 1);
        int count = 1;
        for (int i = 1; i < arrayA.size(); i++) {
            left[i] = left[i-1] * arrayA[i-1];
        }
        for (int i = arrayA.size()-1; i >= 0; i--) {
            left[i] *= count;
            count *= arrayA[i]; 
        }
        return left;
    }
};

// 直接除会有问题，比如0，所以直接左边的乘积乘以右边的乘积
// O(n), O(1)